daemon/cmd/ciliumendpoints.go

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiTypes "k8s.io/apimachinery/pkg/types"
+
+	"github.com/cilium/cilium/pkg/endpoint"
+	cilium_v2a1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	ciliumv2 "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+)
+
+type localEndpointCache interface {
+	LookupPodName(name string) *endpoint.Endpoint
+}
+
+// This must only be run after K8s Pod and CES/CEP caches are synced and local endpoint restoration is complete.
+func (d *Daemon) cleanStaleCEPs(ctx context.Context, eps localEndpointCache, ciliumClient ciliumv2.CiliumV2Interface, enableCiliumEndpointSlice bool) error {
+	crdType := "ciliumendpoint"
+	if enableCiliumEndpointSlice {
+		crdType = "ciliumendpointslice"
+	}
+	indexer := d.k8sWatcher.GetIndexer(crdType)
+	objs, err := indexer.ByIndex("localNode", node.GetCiliumEndpointNodeIP())
+	if err != nil {
+		return fmt.Errorf("could not get %s objects from localNode indexer: %w", crdType, err)
+	}
+	if enableCiliumEndpointSlice {
+		for _, cesObj := range objs {
+			ces, ok := cesObj.(*cilium_v2a1.CiliumEndpointSlice)
+			if !ok {
+				return fmt.Errorf("unexpected object type returned from ciliumendpointslice store: %T", cesObj)
+			}
+			for _, cep := range ces.Endpoints {
+				if cep.Networking.NodeIP == node.GetCiliumEndpointNodeIP() {
+					d.checkCiliumEndpoint(ctx, ces.Namespace, cep.Name, nil, ciliumClient, eps)
+				}
+			}
+		}
+	} else {
+		for _, cepObj := range objs {
+			cep, ok := cepObj.(*types.CiliumEndpoint)
+			if !ok {
+				return fmt.Errorf("unexpected object type returned from ciliumendpoint store: %T", cepObj)
+			}
+
+			if cep.Networking.NodeIP == node.GetCiliumEndpointNodeIP() {
+				d.checkCiliumEndpoint(ctx, cep.Namespace, cep.Name, &cep.ObjectMeta.UID, ciliumClient, eps)
+			}
+		}
+	}
+	return nil
+}
+
+func (d *Daemon) checkCiliumEndpoint(
```

Comment:
The name of this function doesn't just do checking of the CEP, but it also might delete. We should clarify this and add a godoc as well, because it's definitely a non-trivial function.

Reply:
It seems like we can move the `if eps.LookupPodName(cepNamespace+"/"+cepName) == nil {` condition to the caller and then that would allow us to call this function `deleteStaleCEP`.

Reply:
good idea

Reply:
refactored and added better docs